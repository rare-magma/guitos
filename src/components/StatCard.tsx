import { Stat } from "./Stat";
import { Button, Card, Form, InputGroup } from "react-bootstrap";

function formatDescription(description: string): string {
  return description.substring(0, 60) + "...";
}

interface StatCardProps {
  stat: Stat;
  onEdit: (stat: Stat) => void;
}

function StatCard(props: StatCardProps) {
  const { stat, onEdit } = props;

  const handleEditClick = (statBeingEdited: Stat) => {
    onEdit(statBeingEdited);
  };

  return (
    <Card>
      <Card.Header>Stats</Card.Header>
      <Card.Body>
        <InputGroup className="mb-1" key={stat.id + "stats"}>
          <InputGroup.Text>available</InputGroup.Text>
          <Form.Control
            aria-label={"available"}
            key={stat.id + "available"}
            defaultValue={stat.available}
            disabled
            readOnly
          />
        </InputGroup>
        <InputGroup className="mb-1" key={stat.id + "withGoal"}>
          <InputGroup.Text>with goal</InputGroup.Text>
          <Form.Control
            aria-label={"withGoal"}
            key={stat.id + "withGoal"}
            defaultValue={stat.withGoal}
            disabled
            readOnly
          />
        </InputGroup>
        <InputGroup className="mb-1" key={stat.id + "saved"}>
          <InputGroup.Text>saved</InputGroup.Text>
          <Form.Control
            aria-label={"saved"}
            key={stat.id + "saved"}
            defaultValue={stat.saved}
            disabled
            readOnly
          />
        </InputGroup>
        <InputGroup className="mb-1" key={stat.id + "goal"}>
          <InputGroup.Text>goal</InputGroup.Text>
          <Form.Control
            aria-label={"goal"}
            key={stat.id + "goal"}
            defaultValue={stat.goal}
          />
        </InputGroup>
        <InputGroup className="mb-1" key={stat.id + "savings"}>
          <InputGroup.Text>savings</InputGroup.Text>
          <Form.Control
            aria-label={"savings"}
            key={stat.id + "savings"}
            defaultValue={stat.savings}
            disabled
            readOnly
          />
          <Button variant="primary" id="button-addon1">
            +
          </Button>
        </InputGroup>
      </Card.Body>
    </Card>
  );
}

export default StatCard;
