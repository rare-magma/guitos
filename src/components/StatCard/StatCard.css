.auto-goal {
  border: 1px solid var(--comment);
  border-radius: 0.375rem;
}

.progress {
  background-color: var(--comment);
  height: 3px;
}

.progress-bar {
  background-color: var(--pink);
}
