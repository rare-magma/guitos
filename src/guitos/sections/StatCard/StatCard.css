.auto-goal {
  border: 1px solid var(--lightbgcolor);
  border-radius: 0.375rem;
}

.progress {
  background-color: var(--comment);
  height: 3px;
}

.progress-bar {
  background-color: var(--pink);
}
