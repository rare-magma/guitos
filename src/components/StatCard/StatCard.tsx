import { useRef, useState } from "react";
import {
  Button,
  Card,
  Col,
  Form,
  InputGroup,
  OverlayTrigger,
  ProgressBar,
  Row,
  Tooltip,
} from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import { useHotkeys } from "react-hotkeys-hook";

import { produce } from "immer";
import {
  BsArrowReturnRight,
  BsGear,
  BsGraphUp,
  BsPercent,
} from "react-icons/bs";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import {
  calcAutoGoal,
  calcAvailable,
  calcSaved,
  calcWithGoal,
  focusRef,
  parseLocaleNumber,
  roundBig,
} from "../../utils";
import "./StatCard.css";

interface StatCardProps {
  onShowGraphs: () => void;
}

export function StatCard({ onShowGraphs }: StatCardProps) {
  const { intlConfig } = useConfig();
  const { revenuePercentage, budget, setBudget } = useBudget();
  const stat = budget?.stats;
  const [autoGoal, setAutoGoal] = useState(false);

  const shouldCalculateAvailablePerc =
    revenuePercentage <= 100 && stat && stat.available > 0;

  const goalRef =
    useRef<HTMLInputElement>() as React.MutableRefObject<HTMLInputElement>;
  const reservesRef =
    useRef<HTMLInputElement>() as React.MutableRefObject<HTMLInputElement>;

  useHotkeys("g", (e) => !e.repeat && focusRef(goalRef), {
    preventDefault: true,
  });
  useHotkeys("e", (e) => !e.repeat && focusRef(reservesRef), {
    preventDefault: true,
  });

  function handleInputChange(item: React.ChangeEvent<HTMLInputElement>) {
    if (budget && item) {
      setAutoGoal(false);
      const newState = produce((draft) => {
        draft.stats.goal = item.target.valueAsNumber;
        draft.stats.available = roundBig(calcAvailable(draft), 2);
        draft.stats.withGoal = calcWithGoal(draft);
        draft.stats.saved = calcSaved(draft);
      }, budget);
      setBudget(newState(), false);
    }
  }

  function handleReserveChange(value: string | undefined): void {
    if (budget && value) {
      const newState = produce((draft) => {
        draft.stats.reserves = parseLocaleNumber(value, intlConfig?.locale);
      }, budget);
      setBudget(newState(), false);
    }
  }

  function handleAutoGoal() {
    if (budget && stat) {
      const newState = produce((draft) => {
        draft.stats.goal = calcAutoGoal(draft);
        draft.stats.available = roundBig(calcAvailable(draft), 2);
        draft.stats.withGoal = calcWithGoal(draft);
        draft.stats.saved = calcSaved(draft);
      }, budget);
      setBudget(newState(), true);
    }
    setAutoGoal(true);
  }

  return (
    <Card
      className="stat-card"
      key={`stat-${intlConfig?.currency}-${budget?.expenses.total} + ${budget?.incomes.total}}`}
    >
      <Card.Header className="stat-card-header py-0">
        <Row className="mb-1">
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip
                id={`tooltip-value-statistics-bar`}
                style={{ position: "fixed" }}
              >
                {revenuePercentage}% of revenue spent
              </Tooltip>
            }
          >
            <Col className="align-self-center">
              Statistics
              <ProgressBar
                role="progressbar"
                aria-label="percentage of revenue spent"
                aria-valuetext={`${revenuePercentage}%`}
                min={0}
                max={100}
                now={revenuePercentage}
                label={`${revenuePercentage}%`}
                visuallyHidden
              />
            </Col>
          </OverlayTrigger>
          <Col xs="auto" className="text-end">
            <OverlayTrigger
              placement="top"
              overlay={
                <Tooltip id={`tooltip-graphs`} style={{ position: "fixed" }}>
                  open charts view
                </Tooltip>
              }
            >
              <Button
                aria-label="open charts view"
                className="my-1"
                size="sm"
                variant="outline-primary"
                style={{ color: "var(--textcolor)" }}
                onClick={() => onShowGraphs()}
              >
                <BsGraphUp aria-hidden />
              </Button>
            </OverlayTrigger>
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        <InputGroup className="mb-1" key={"stats"}>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={`tooltip-available`} style={{ position: "fixed" }}>
                = revenue - expenses
              </Tooltip>
            }
          >
            <InputGroup.Text>available</InputGroup.Text>
          </OverlayTrigger>
          <CurrencyInput
            id="available"
            key={`${stat?.available}-available`}
            className="text-end form-control fixed-width-font"
            aria-label={"available"}
            name="available"
            intlConfig={intlConfig}
            disabled={true}
            defaultValue={stat?.available}
          />
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip
                id={`tooltip-value-available-perc`}
                style={{ position: "fixed" }}
              >
                % of revenue available
              </Tooltip>
            }
          >
            <InputGroup.Text aria-label="percentage of revenue available">
              {shouldCalculateAvailablePerc ? 100 - revenuePercentage : 0}
              <BsPercent aria-hidden />
            </InputGroup.Text>
          </OverlayTrigger>
        </InputGroup>
        <InputGroup className="mb-1" key={"withGoal"}>
          <InputGroup.Text>
            <BsArrowReturnRight aria-hidden />
            <OverlayTrigger
              placement="top"
              overlay={
                <Tooltip id={`tooltip-available`} style={{ position: "fixed" }}>
                  available with goal = available - savings estimate
                </Tooltip>
              }
            >
              <div style={{ marginLeft: ".5rem" }}>with goal</div>
            </OverlayTrigger>
          </InputGroup.Text>
          <CurrencyInput
            id="with-goal"
            key={`${stat?.withGoal}-withGoal`}
            className="text-end form-control fixed-width-font"
            aria-label={"with-goal"}
            name="with-goal"
            intlConfig={intlConfig}
            disabled={true}
            defaultValue={stat?.withGoal}
          />
        </InputGroup>
        <InputGroup className="mb-1" key={"goal"}>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={`tooltip-available`} style={{ position: "fixed" }}>
                % of revenue that should go into savings
              </Tooltip>
            }
          >
            <InputGroup.Text>savings goal</InputGroup.Text>
          </OverlayTrigger>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={`tooltip-auto-goal`} style={{ position: "fixed" }}>
                estimate savings goal
              </Tooltip>
            }
          >
            <Button
              aria-label="calculate savings goal"
              variant="outline-primary"
              className="input-group-text auto-goal"
              style={{ color: "var(--textcolor)" }}
              onClick={handleAutoGoal}
            >
              <BsGear aria-hidden />
            </Button>
          </OverlayTrigger>
          <Form.Control
            id="goal-input"
            data-testid="goal-input"
            className="text-end fixed-width-font"
            aria-label={"goal"}
            key={`auto-goal-${autoGoal}`}
            defaultValue={stat?.goal}
            onChange={handleInputChange}
            onWheel={(e) => e.target instanceof HTMLElement && e.target.blur()}
            type="number"
            ref={goalRef}
            onInput={(e: React.ChangeEvent<HTMLInputElement>) => {
              e.target.valueAsNumber = Number(
                Math.max(0, Number(e.target.value)).toString().slice(0, 2),
              );
            }}
          />
          <InputGroup.Text>
            <BsPercent aria-hidden />
          </InputGroup.Text>
        </InputGroup>
        <InputGroup className="mb-1" key={"saved"}>
          <InputGroup.Text>
            <BsArrowReturnRight aria-hidden />
            <OverlayTrigger
              placement="top"
              overlay={
                <Tooltip id={`tooltip-available`} style={{ position: "fixed" }}>
                  savings estimate = savings goal * available / 100
                </Tooltip>
              }
            >
              <div style={{ marginLeft: ".5rem" }}>estimate</div>
            </OverlayTrigger>
          </InputGroup.Text>
          <CurrencyInput
            id="saved"
            key={`${stat?.saved}-saved`}
            className="text-end form-control fixed-width-font"
            aria-label={"saved"}
            name="saved"
            intlConfig={intlConfig}
            disabled={true}
            defaultValue={stat?.saved}
          />
        </InputGroup>
        <InputGroup className="mb-1" key={"reserves"}>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={`tooltip-available`} style={{ position: "fixed" }}>
                emergency fund/cash
              </Tooltip>
            }
          >
            <InputGroup.Text>reserves</InputGroup.Text>
          </OverlayTrigger>
          <CurrencyInput
            id="reserves"
            key={"reserves-input"}
            className="text-end form-control fixed-width-font"
            aria-label={"reserves"}
            name="reserves"
            intlConfig={intlConfig}
            defaultValue={stat?.reserves}
            maxLength={14}
            allowNegativeValue={false}
            ref={reservesRef}
            onFocus={() => {
              reservesRef.current?.setSelectionRange(0, 25);
            }}
            onValueChange={(value) => handleReserveChange(value)}
          />
        </InputGroup>
      </Card.Body>
    </Card>
  );
}

export default StatCard;
