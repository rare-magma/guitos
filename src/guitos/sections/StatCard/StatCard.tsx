import { produce } from "immer";
import type React from "react";
import { useEffect, useRef, useState } from "react";
import {
  Button,
  Card,
  Col,
  Form,
  InputGroup,
  OverlayTrigger,
  ProgressBar,
  Row,
  Tooltip,
} from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import { useHotkeys } from "react-hotkeys-hook";
import {
  BsArrowReturnRight,
  BsGear,
  BsGraphUp,
  BsPercent,
} from "react-icons/bs";
import "./StatCard.css";
import { BudgetCalculator } from "@guitos/application/budgetCalculator";
import { useBudget } from "@guitos/context/BudgetContext";
import { useConfig } from "@guitos/context/ConfigContext";
import type { Budget } from "@guitos/domain/budget";
import { focusRef, parseLocaleNumber, roundBig } from "../../../utils";

interface StatCardProps {
  onShowGraphs: () => void;
}

export function StatCard({ onShowGraphs }: StatCardProps) {
  const { userOptions, intlConfig } = useConfig();
  const { revenuePercentage, budget, setBudget } = useBudget();
  const stat = budget?.stats;
  const [autoGoal, setAutoGoal] = useState(false);
  const progressBarRef = useRef<HTMLDivElement>(null);

  const shouldCalculateAvailablePerc =
    revenuePercentage <= 100 && stat && stat.available > 0;

  const goalRef = useRef<HTMLInputElement>(null);
  const reservesRef = useRef<HTMLInputElement>(null);

  useHotkeys("g", (e) => !e.repeat && focusRef(goalRef), {
    preventDefault: true,
  });
  useHotkeys("e", (e) => !e.repeat && focusRef(reservesRef), {
    preventDefault: true,
  });

  function handleInputChange(item: React.ChangeEvent<HTMLInputElement>) {
    if (budget && item) {
      setAutoGoal(false);
      const newState = produce((draft) => {
        draft.stats.goal = item.target.valueAsNumber;
        draft.stats.available = roundBig(
          BudgetCalculator.available(draft as Budget),
          2,
        );
        draft.stats.withGoal = BudgetCalculator.availableWithGoal(
          draft as Budget,
        );
        draft.stats.saved = BudgetCalculator.saved(draft as Budget);
      }, budget);
      setBudget(newState(), false);
    }
  }

  function handleReserveChange(value: string | undefined): void {
    if (budget && value) {
      const newState = produce((draft) => {
        draft.stats.reserves = parseLocaleNumber(value, userOptions.locale);
      }, budget);
      setBudget(newState(), false);
    }
  }

  function handleAutoGoal() {
    if (budget && stat) {
      const newState = produce((draft) => {
        draft.stats.goal = BudgetCalculator.automaticGoal(draft as Budget);
        draft.stats.available = roundBig(
          BudgetCalculator.available(draft as Budget),
          2,
        );
        draft.stats.withGoal = BudgetCalculator.availableWithGoal(
          draft as Budget,
        );
        draft.stats.saved = BudgetCalculator.saved(draft as Budget);
      }, budget);
      setBudget(newState(), true);
    }
    setAutoGoal(true);
  }

  // workaround for https://github.com/react-bootstrap/react-bootstrap/pull/6739
  useEffect(() => {
    const innerBar = progressBarRef.current?.querySelector(".progress-bar");
    if (innerBar) {
      innerBar.setAttribute("aria-label", "percentage of revenue spent");
    }
  }, []);

  return (
    <Card
      className="stat-card"
      key={`stat-${userOptions.currencyCode}-${budget?.expenses.total} + ${budget?.incomes.total}}`}
    >
      <Card.Header className="stat-card-header py-0">
        <Row className="mb-1">
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip
                id={"tooltip-value-statistics-bar"}
                style={{ position: "fixed" }}
              >
                {revenuePercentage}% of revenue spent
              </Tooltip>
            }
          >
            <Col className="align-self-center">
              Statistics
              <ProgressBar
                ref={progressBarRef}
                role="progressbar"
                aria-label="percentage of revenue spent"
                aria-valuetext={`${revenuePercentage}%`}
                min={0}
                max={100}
                now={revenuePercentage}
                visuallyHidden={true}
              />
            </Col>
          </OverlayTrigger>
          <Col xs="auto" className="text-end">
            <OverlayTrigger
              placement="top"
              overlay={
                <Tooltip id={"tooltip-graphs"} style={{ position: "fixed" }}>
                  open charts view
                </Tooltip>
              }
            >
              <Button
                aria-label="open charts view"
                className="my-1"
                size="sm"
                variant="outline-primary"
                style={{ color: "var(--textcolor)" }}
                onClick={() => onShowGraphs()}
              >
                <BsGraphUp aria-hidden={true} />
              </Button>
            </OverlayTrigger>
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        <InputGroup className="mb-1" key={"stats"}>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={"tooltip-available"} style={{ position: "fixed" }}>
                = revenue - expenses
              </Tooltip>
            }
          >
            <InputGroup.Text>available</InputGroup.Text>
          </OverlayTrigger>
          <CurrencyInput
            key={`${stat?.available}-available`}
            className="text-end form-control fixed-width-font"
            aria-label={"available"}
            name="available"
            intlConfig={intlConfig}
            disabled={true}
            defaultValue={stat?.available}
          />
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip
                id={"tooltip-value-available-perc"}
                style={{ position: "fixed" }}
              >
                % of revenue available
              </Tooltip>
            }
          >
            <InputGroup.Text aria-label="percentage of revenue available">
              {shouldCalculateAvailablePerc ? 100 - revenuePercentage : 0}
              <BsPercent aria-hidden={true} />
            </InputGroup.Text>
          </OverlayTrigger>
        </InputGroup>
        <InputGroup className="mb-1" key={"withGoal"}>
          <InputGroup.Text>
            <BsArrowReturnRight aria-hidden={true} />
            <OverlayTrigger
              placement="top"
              overlay={
                <Tooltip id={"tooltip-available"} style={{ position: "fixed" }}>
                  available with goal = available - savings estimate
                </Tooltip>
              }
            >
              <div style={{ marginLeft: ".5rem" }}>with goal</div>
            </OverlayTrigger>
          </InputGroup.Text>
          <CurrencyInput
            key={`${stat?.withGoal}-withGoal`}
            className="text-end form-control fixed-width-font"
            aria-label={"with-goal"}
            name="with-goal"
            intlConfig={intlConfig}
            disabled={true}
            defaultValue={stat?.withGoal}
          />
        </InputGroup>
        <InputGroup className="mb-1" key={"goal"}>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={"tooltip-available"} style={{ position: "fixed" }}>
                % of revenue that should go into savings
              </Tooltip>
            }
          >
            <InputGroup.Text>savings goal</InputGroup.Text>
          </OverlayTrigger>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={"tooltip-auto-goal"} style={{ position: "fixed" }}>
                estimate savings goal
              </Tooltip>
            }
          >
            <Button
              aria-label="calculate savings goal"
              variant="outline-primary"
              className="input-group-text auto-goal"
              style={{ color: "var(--textcolor)" }}
              onClick={handleAutoGoal}
            >
              <BsGear aria-hidden={true} />
            </Button>
          </OverlayTrigger>
          <Form.Control
            data-testid="goal-input"
            className="text-end fixed-width-font"
            aria-label={"goal"}
            key={`auto-goal-${autoGoal}`}
            defaultValue={stat?.goal}
            onChange={handleInputChange}
            onWheel={(e) => e.target instanceof HTMLElement && e.target.blur()}
            type="number"
            ref={goalRef}
            onInput={(e: React.ChangeEvent<HTMLInputElement>) => {
              e.target.valueAsNumber = Number(
                Math.max(0, Number(e.target.value)).toString().slice(0, 2),
              );
            }}
          />
          <InputGroup.Text>
            <BsPercent aria-hidden={true} />
          </InputGroup.Text>
        </InputGroup>
        <InputGroup className="mb-1" key={"saved"}>
          <InputGroup.Text>
            <BsArrowReturnRight aria-hidden={true} />
            <OverlayTrigger
              placement="top"
              overlay={
                <Tooltip id={"tooltip-available"} style={{ position: "fixed" }}>
                  savings estimate = savings goal * available / 100
                </Tooltip>
              }
            >
              <div style={{ marginLeft: ".5rem" }}>estimate</div>
            </OverlayTrigger>
          </InputGroup.Text>
          <CurrencyInput
            key={`${stat?.saved}-saved`}
            className="text-end form-control fixed-width-font"
            aria-label={"saved"}
            name="saved"
            intlConfig={intlConfig}
            disabled={true}
            defaultValue={stat?.saved}
          />
        </InputGroup>
        <InputGroup className="mb-1" key={"reserves"}>
          <OverlayTrigger
            placement="top"
            overlay={
              <Tooltip id={"tooltip-available"} style={{ position: "fixed" }}>
                emergency fund/cash
              </Tooltip>
            }
          >
            <InputGroup.Text>reserves</InputGroup.Text>
          </OverlayTrigger>
          <CurrencyInput
            key={"reserves-input"}
            className="text-end form-control fixed-width-font"
            aria-label={"reserves"}
            name="reserves"
            intlConfig={intlConfig}
            defaultValue={stat?.reserves}
            maxLength={14}
            allowNegativeValue={false}
            ref={reservesRef}
            onFocus={() => {
              reservesRef.current?.setSelectionRange(0, 25);
            }}
            onValueChange={(value) => handleReserveChange(value)}
          />
        </InputGroup>
      </Card.Body>
    </Card>
  );
}
