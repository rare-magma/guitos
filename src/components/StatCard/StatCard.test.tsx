import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import StatCard from "./StatCard";

describe("StatCard", () => {
  const onChange = vi.fn();
  const onAutoGoal = vi.fn();
  const onShowGraphs = vi.fn();
  const comp = (
    <StatCard
      onChange={onChange}
      onAutoGoal={onAutoGoal}
      onShowGraphs={onShowGraphs}
    />
  );

  beforeEach(() => {
    render(comp);
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    expect(screen.getByText("Statistics")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$90")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$80")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$10")).toBeInTheDocument();
    expect(screen.getByDisplayValue("10")).toBeInTheDocument();
  });

  it("triggers onChange when user changes input", async () => {
    await userEvent.type(screen.getByLabelText("reserves"), "2");

    expect(onChange).toBeCalledWith({
      available: 90,
      goal: 10,
      reserves: 2,
      saved: 10,
      withGoal: 80,
    });
    expect(screen.getByDisplayValue("$2")).toBeInTheDocument();

    await userEvent.clear(screen.getByTestId("goal-input"));
    await userEvent.type(screen.getByTestId("goal-input"), "95");

    expect(onChange).toBeCalledWith({
      available: 90,
      goal: 95,
      reserves: 2,
      saved: 10,
      withGoal: 80,
    });
    expect(screen.getByDisplayValue("95")).toBeInTheDocument();
  });

  it("triggers onAutoGoal when user clicks button", async () => {
    await userEvent.click(
      screen.getByRole("button", { name: "calculate savings goal" }),
    );

    expect(onAutoGoal).toBeCalledWith({
      available: 90,
      goal: 95,
      reserves: 2,
      saved: 10,
      withGoal: 80,
    });
  });

  it("triggers onShowGraphs when user clicks button", async () => {
    await userEvent.click(
      screen.getByRole("button", { name: "open charts view" }),
    );

    expect(onShowGraphs).toBeCalledTimes(1);
  });
});
