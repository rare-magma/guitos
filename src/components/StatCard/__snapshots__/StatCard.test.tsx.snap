// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatCard > matches snapshot 1`] = `
<StatCard
  onAutoGoal={[MockFunction spy]}
  onChange={[MockFunction spy]}
  onShowGraphs={[MockFunction spy]}
/>
`;
